### DISCUSSION TITLE
Hang occuring when multiple `clangd` servers are started on desktop lazy load.

### DISCUSSION THREAD

**Discussion by sonictk** [joaotavora/eglot#970] [Q&A]
I'm encountering a really odd issue that has me stumped.

When I use `eglot` with `desktop-read`, I'm encountering situations where Emacs freezes after starting `clangd` and eglot tries to connect to it. It doesn't seem to be tied to a specific number of servers, (sometimes it happens even when just 2 servers are started). It seems like the Emacs process is forever waiting for one of the `clangd.exe` processes to be signaled from the wait state when I attach a debugger to `Emacs.exe`. 

I've tried deleting everything from my startup `.el` file to just the following:

```
; Add custom module path so that nothing is saved to the global emacs config
(add-to-list 'load-path "~/Git/lightweight-emacs/modules/")

(require 'eglot)
(add-to-list 'eglot-server-programs '((c++-mode c-mode objc-mode cuda-mode) "clangd"))
(add-hook 'c-mode-hook 'eglot-ensure)
(add-hook 'c++-mode-hook 'eglot-ensure)

(setq eglot-autoshutdown t)
(setq eglot-autoreconnect nil)
(setq eglot-connect-timeout 10)
(setq eglot-strict-mode nil)
(setq eglot-extend-to-xref t)

; Save the Emacs sessions
(setq desktop-dirname             "~/Git/lightweight-emacs/"
      desktop-base-file-name      "emacs.desktop"
      desktop-base-lock-name      "lock"
      desktop-path                (list desktop-dirname)
      desktop-save                t
      desktop-restore-eager       8 ; Number of buffers to restore immediately; rest are lazily loaded when emacs is idle
      desktop-lazy-verbose        t
      desktop-lazy-idle-delay     5 
      desktop-files-not-to-save   "^$" ;reload tramp paths
      desktop-load-locked-desktop t
      desktop-auto-save-timeout   30)
(desktop-save-mode 1)
```

And I can pretty much reproduce this on-demand when working in an Unreal Engine project (which has a pretty large `compile_commands.json` database, about 700+ MB). The buffers usually restore lazily fine, but the moment I give Emacs focus, it freezes. The only solution is to kill `clangd.exe` processes one-by-one until the "right" one that is locking up Emacs is found, and Emacs will again become responsive.

It's worth noting that this sometimes happens outside of just desktop lazy loading; literally, I'll open a new buffer, eglot will try to launch, and immediately hang Emacs.

For some weird reason, this doesn't happen with `lsp-mode` with the exact same version of `clangd`, so I'm inclined to believe this is something eglot-specific. I don't know why, though.

There are no notable error messages in the *Messages* buffer other than:

```
Lazy desktop load complete
[eglot] Connected! Server `clangd' now managing `c++-mode' buffers in project `Containers'.
[eglot] Connected! Server `clangd' now managing `c++-mode' buffers in project `Public'.
```

And the eglot events buffer doesn't have anything either:

```
[stderr] I[22:45:56.289] clangd version 14.0.4
[stderr] I[22:45:56.289] Features: windows
[stderr] I[22:45:56.289] PID: 17220
[stderr] I[22:45:56.289] Working directory: s:\source\repos\epic\ysiew_fn_release_21_00\Engine\Restricted\NotForLicensees\Plugins\Solaris\Source\SolarisEditor\Private
[stderr] I[22:45:56.289] argv[0]: c:\Program Files\LLVM\bin\clangd.exe
[stderr] I[22:45:56.297] Starting LSP over stdin/stdout
[stderr] I[22:46:04.282] <-- initialize(1)
[internal] Wed Jun  8 22:46:07 2022:
(:message "Connection state changed" :change "exited abnormally with code 1\n")

----------b---y---e---b---y---e----------
[stderr] 
[stderr] 
[stderr] nil
[stderr] nil
[stderr] Process EGLOT (Private/c++-mode) stderr finished
[internal] Wed Jun  8 22:46:09 2022:
(:message "Connection state changed" :change "exited abnormally with code 1\n")

----------b---y---e---b---y---e----------
[internal] Wed Jun  8 22:46:09 2022:
(:message "Connection state changed" :change "exited abnormally with code 1\n")

----------b---y---e---b---y---e----------
```
Which only show the messages about me force-killing `clangd.exe` in order to get Emacs responsive again.

This is just really weird and has been bugging me for months now. Does anyone else have any ideas about what I can try to debug what's going on? I downloaded a debug build of Emacs and tried attaching the debugger while it was in the hung state but the debug build didn't seem to have symbols available so I couldn't see any callstacks either.

Again, to summarize:

- It feels like somehow when more than one clangd server is started, Emacs can somehow hang. It's as if one clangd server is blocking Emacs from starting the other, somehow. 
- This happens regardless of desktop lazy loading (i.e. if I visit a buffer manually and call `eglot` to start the 2nd server, it happens immediately.)
- Only way to unfreeze emacs is to kill clangd processes one-by-one until it becomes responsive again.

As an addendum: when this happens (which is basically almost every single time I start more than one `clangd` server), and I attach a debugger to Emacs, this is what I see:

![image](https://user-images.githubusercontent.com/2504199/174424703-a3402e70-6949-4e37-b6f4-5420d88f940c.png)

From what I can understand (I couldn't find any information on `_lock_fhandle` even on MSDN, but it appears to be part of the older VCRT), it seems like the reason for Emacs hanging is that it tries to acquire the mutex for a file somehow after a `clangd` process has been launched, but one of the `clangd` processes is not giving up the lock for it. Is there something in `eglot.el` that perhaps indirectly asks Emacs to do this? Even then, I'm not sure if there's anything I can do here since I'm not sure why Emacs is trying to do this. 

If I try running this with Emacs 29, I get an error that says `error in process sentinel: Creating pipe: Too many open files`. Is there a way to hard-limit the number of files that `clangd` tries to read from?

Any ideas for how I could get more info to resolve this would be much appreciated! I've been trying to find a way if I can just get PDB files for Emacs builds on Windows but it doesn't seem like this is trivial.

**Comment by joaotavora** [comment]
Sorry for the delay here.  This is too complicated to analyze right now, but it's on the radar, just wanted to let you know.  I'll skim your latest post and see if anything obvious pops out to me.

**Reply by joaotavora** [reply]
> I'll skim your latest post and see if anything obvious pops out to me.

It _could_ be that Eglot suddently has sent the server very many `textDocument/didOpen`all at once.  But these can't be seen in the events buffer.  How many such buffers do you have open, or how many does `desktop-read` try to create?

> debug build didn't seem to have symbols available so I couldn't see any callstacks either.

A debug build normally has debug symbols.  Are you looking in the correct thread?  I'm not familiar with the debugger you presented, I just know gdb.

I also feel you should try skimming your recipe even further.  And many summarize the problem again.

Can you start the server manually not as a subprocess of Emacs, but simply as a process of a shell?  Can you give it input via stdin, similar to what Eglot's input would be?  Does all of this work when `desktop-read` is out of the picture?

My understanding right now is that the breaking combination is something like this:

* `eglot`
* `desktop-read`
* very many files open, or waiting to be opened or something
* `clangd`
* Windows

My understanding of what exactly hangs in shakier yet.  Does Emacs hang?  Does clangd hang?

**Reply by sonictk** [reply]
Hey, thanks for reading. I know I might have typed up the problem a little bit too verbose, sorry about that. I'll try to summarize here:

> My understanding of what exactly hangs in shakier yet. Does Emacs hang? Does clangd hang?

- `emacs.exe` itself hangs, along with one of the `clangd` processes, with me being unable to send Emacs input or do anything (i.e. ctrl-g doesn't work). If you see the Parallel Stacks window (this is from Visual Studio) screenshot above when the debugger is attached, when this happens, the other `clangd.exe` processes are running fine, but one of the `clangd.exe` processes is always stuck in the middle of reading a file, while Emacs itself is stuck in the middle of closing a file handle, it looks like. This is where I really wish I had the PDBs for Emacs so that the debugger could give some named symbols in the callstacks.

> It could be that Eglot suddently has sent the server very many textDocument/didOpenall at once. But these can't be seen in the events buffer. How many such buffers do you have open, or how many does desktop-read try to create?

It doesn't appear to be directly related to `desktop-read` (because I can repro this sometimes just by opening a new buffer and running `eglot` (so it's not related to having `(add-hook 'c++-mode-hook 'eglot-ensure)` either). I apologize for the confusion; what I meant to say was that `desktop-read` can easily trigger the issue since it can try to restore several buffers at once.

As for how many buffers I have open at once: usually I have just 5 - 10, but I can repro this with just 2 open, and it does not repro consistently (i.e. I can have the same 2 buffers and hit the issue once, kill Emacs, restart Emacs, open the same 2 buffers again and start `eglot` in each of them in the same order and it will not happen the 2nd time round.)

> I also feel you should try skimming your recipe even further.

I've tried trimming my emacs init file to be literally just:

```
(add-to-list 'load-path "~/Git/lightweight-emacs/modules/")
(require 'eglot)
(add-to-list 'eglot-server-programs '((c++-mode c-mode objc-mode cuda-mode) "clangd"))
```

And nothing else, and I can still repro this issue.

I've tried:

- disabling `clangd` indexing globally to see if it was the issue. No use.
- disabling `desktop-read`'s lazy loading and restoring all buffers at once. Also didn't work.
- all sorts of random tweaks like changing the connection timeout, eglot's event buffer size, etc. with no success.

The only thing that works when this happens is that I have to close `clangd.exe` processes, one by one, until I find the one that is hung (and that Emacs is apparently waiting for to release a file lock or something) before Emacs will suddenly become responsive again. The confusing thing is that, when I restart the server I just killed at this point, the problem sometimes doesn't occur again and the server successfully runs fine. Then I'll go open another buffer, and then suddenly the problem will manifest again with a totally different server.

> A debug build normally has debug symbols. Are you looking in the correct thread? I'm not familiar with the debugger you presented, I just know gdb.

I tried downloading the debug build of Emacs from http://mirror.us-midwest-1.nexcess.net/gnu/emacs/windows/emacs-28/ , but it doesn't seem to come with `.pdb` files, which contain the debug symbols needed for debuggers on Windows to be able to identify functions. I think it's just an optimized build that was made without symbols. If there are PDBs available for Emacs on Windows builds, or if there's a good way to build Emacs using MSVC I would be glad to try and build it from source to figure this out.

> Can you start the server manually not as a subprocess of Emacs, but simply as a process of a shell? Can you give it input via stdin, similar to what Eglot's input would be? Does all of this work when desktop-read is out of the picture?

This is a good idea, thank you; I'll try this out. Can I start the `clangd` server from a separate terminal and have it communicate with `eglot`? I've been having trouble figuring out how to do that.

I wish I could share the `compile_commands.json` file I'm using, but unfortunately I cannot. I suspect this is somehow related to the fact that this `compile_commands.json` file is large, that is triggering something with how `eglot` launches `clangd` within Emacs, because I have never experienced this issue with smaller `compile_commands.json` files. 

Thanks for the response BTW, I really appreciate it. If you can think of anything at all for me to try, I'd be happy to give it a go.

**Reply by joaotavora** [reply]
> I meant to say was that desktop-read can easily trigger the issue since it can try to restore several buffers at once.

Ah, ok that makes sense.   How often would you say the problem happens witout `desktop-read`.  If it happens often (for some reasonable meaning of "often") then you can issue from a shell

```
path/to/emacs.exe -Q -L path/to/eglot.el -l eglot  path/to/a/c/file.c -f eglot -f kill-emacs
```

and then see how often you can call it until it hangs.  With a debug-compiler emacs.exe and some gdb and -- importantly -- with the help of Emacs/W32 specialist Eli Zaretskii we can get to the bottom of the issue.

> I wish I could share the compile_commands.json file I'm using, but unfortunately I cannot. I suspect this is somehow related to the fact that this compile_commands.json file is large, that is triggering something with how eglot launches clangd within Emacs, because I have never experienced this issue with smaller compile_commands.json files.

In that case, it would point to some delay on clangd.exe's part and to an inability of Eglot to deal with such a delay.  In fact, now that you mention it, it seems that Eglot/Emacs is so hung in that conversation that it never sends the `initialize` message:
```
[stderr] I[22:45:56.297] Starting LSP over stdin/stdout
[stderr] I[22:46:04.282] <-- initialize(1)
```

Or maybe it does send it, but never prints it (because it is hung).

**Comment by sonictk** [comment]
Ok, one thing I finally did is to just build Emacs from source (`GNU Emacs 29.0.90 (build 3, x86_64-w64-mingw32) of 2023-05-01`) and this issue no longer presents itself. I don't know how or why, but I suspect the new async compilation features have something to do with it.

**Comment by joaotavora** [comment]
On Tue, May 2, 2023 at 7:14 PM Siew Yi Liang ***@***.***>
wrote:

> Ok, one thing I finally did is to just build Emacs from source (GNU Emacs
> 29.0.90 (build 3, x86_64-w64-mingw32) of 2023-05-01) and this issue no
> longer presents itself. I don't know how or why, but I suspect the new
> async compilation features have something to do with it.
>

I don't know what you mean by "async compilation features", but it
seems really odd to me that something

Maybe you were just using an old version of Eglot and/or
missing this commit:

commit 6c8aee268d81ef616169d79ea5bd0331aebc25ee
Author: João Távora ***@***.***>
Date:   Thu Jul 7 12:30:03 2022 +0100

    Prevent desktop.el from saving/restoring eglot--managed-mode

    Although desktop.el compatibility is Emacs bughttps://
github.com/joaotavora/eglot/issues/56407, the optimal
    solution agreed to there is a bit more work than what I have time to
    right now.  See e.g.
    https://debbugs.gnu.org/cgi/bugreport.cgi?bug=bug%2356407

    https://github.com/joaotavora/eglot/issues/68.  For
    now, just use `with-eval-after-load'

    * eglot.el (Hacks desktop): Add eglot--managed-mode to
    desktop-minor-mode-handlers

    GitHub-reference: fix https://github.com/joaotavora/eglot/issues/990

**Comment by joaotavora** [comment]
Or maybe not: I see that you also reproduced this without desktop.el

On Tue, May 2, 2023 at 9:26 PM João Távora ***@***.***> wrote:
>
> On Tue, May 2, 2023 at 7:14 PM Siew Yi Liang ***@***.***> wrote:
>>
>> Ok, one thing I finally did is to just build Emacs from source (GNU Emacs 29.0.90 (build 3, x86_64-w64-mingw32) of 2023-05-01) and this issue no longer presents itself. I don't know how or why, but I suspect the new async compilation features have something to do with it.
>
>
> I don't know what you mean by "async compilation features", but it
> seems really odd to me that something
>
> Maybe you were just using an old version of Eglot and/or
> missing this commit:
>
> commit 6c8aee268d81ef616169d79ea5bd0331aebc25ee
> Author: João Távora ***@***.***>
> Date:   Thu Jul 7 12:30:03 2022 +0100
>
>     Prevent desktop.el from saving/restoring eglot--managed-mode
>
>     Although desktop.el compatibility is Emacs bughttps://github.com/joaotavora/eglot/issues/56407, the optimal
>     solution agreed to there is a bit more work than what I have time to
>     right now.  See e.g.
>     https://debbugs.gnu.org/cgi/bugreport.cgi?bug=bug%2356407
>
>     https://github.com/joaotavora/eglot/issues/68.  For
>     now, just use `with-eval-after-load'
>
>     * eglot.el (Hacks desktop): Add eglot--managed-mode to
>     desktop-minor-mode-handlers
>
>     GitHub-reference: fix https://github.com/joaotavora/eglot/issues/990
>



-- 
João Távora
